public/js/components/Sources.js

```diff
@@ -3,68 +3,261 @@
-  return selectedSource && selectedSource.get("actor") == source.get("actor");
+function nodeHasChildren(item) {
+  return item[2] instanceof Array;
 }
 
-function renderSource({ source, selectSource, selectedSource }) {
-  const pathname = source.get("pathname");
-  const selectedClass = isSelected(selectedSource, source) ? "selected" : "";
+function nodeName(item) {
+  return item[0];
+}
 
-  return dom.li(
-    { onClick: () => selectSource(source.toJS()),
-      className: `source-item ${selectedClass}`,
-      style: { paddingLeft: "20px" },
-      key: source.get("url") },
+function nodePath(item) {
+  return item[1];
+}
 
-    Isvg({ src: "images/angle-brackets.svg" }),
-    dom.span({ className: "label" }, pathname)
-  );
+function nodeContents(item) {
+  return item[2];
 }
 
-/**
- * Takes a sources object indexed by actor and
- * returns a sources object indexed by source domain.
- *
- * @returns Object
- */
-function groupSourcesByDomain(sources) {
-  return sources.valueSeq()
-    .filter(source => !!source.get("url"))
-    .groupBy(source => (new URL(source.get("url"))).hostname);
+function setNodeContents(item, contents) {
+  item[2] = contents;
 }
 
-function Sources({ sources, selectSource, selectedSource }) {
-  const sourcesByDomain = groupSourcesByDomain(sources);
-
-  return dom.div({ className: "sources-panel" },
-    dom.div(
-      { className: "sources-header" }
-    ),
-    dom.ul(
-      { className: "sources-list" },
-      sourcesByDomain.keySeq().map((domain) => {
-        return dom.li({ key: domain, className: "domain" },
-          Isvg({ src: "images/globe.svg" }),
-          dom.span({ className: "label" }, domain),
-          dom.ul(null,
-            sourcesByDomain.get(domain).map(source => renderSource({
-              source, selectSource, selectedSource }))
-          )
-        );
-      })
+function createNode(name, path, contents) {
+  return [name, path, contents];
+}
+
+function createParentMap(tree) {
+  const map = new WeakMap();
+
+  function _traverse(subtree) {
+    if (nodeHasChildren(subtree)) {
+      for (let child of nodeContents(subtree)) {
+        map.set(child, subtree);
+        _traverse(child);
+      }
+    }
+  }
+
+  // Don't link each top-level path to the "root" node because the
+  // user never sees the root
+  nodeContents(tree).forEach(_traverse);
+  return map;
+}
+
+function getURL(source) {
+  try {
+    if (!source.get("url")) {
+      return null;
+    }
+
+    const url = new URL(source.get("url"));
+
+    // Filter out things like `javascript:<code>` URLs for now.
+    // Whitelist the protocols because there may be several strange
+    // ones.
+    if (url.protocol !== "http:" && url.protocol !== "https:") {
+      return null;
+    }
+
+    return url;
+  } catch (e) {
+    // If there is a parse error (which may happen with various
+    // internal script that don't have a correct URL), just ignore it.
+    return null;
+  }
```

Comment:
I hate having to do 

```
let foo;
try { 
  foo = doSomething()
} catch(e) { ... }
```

But on the other hand it's worse to hide all errors in unrelated code like my code does. I think I added code there quickly without thinking about it. Wish JS had better error handling... (is it clear that there are things about JS I don't like? 😛 )